libinteract/libinteract.py

```diff
@@ -867,6 +868,144 @@ def calc_cg_fullmatrix(identifiers, idxs, percmat, perco):
+    table_dict = {"all": table_rows}
+    # Find unique chains in the table
+    chains = np.unique(np.concatenate((table_cols[0], table_cols[4])))
+    # If multiple chains are present, split the contacts by chain
+    if len(chains) > 1:
+	# For each chain, add the nodes that are only in contact with the same chain
+        for chain in chains:
+            filtered_rows_intra = filter_by_chain(chain, chain,
+                                                  table_rows)
+            # Check rows exist
+            if filtered_rows_intra is not None:
+                table_dict[chain] = filtered_rows_intra
+        # Create vector of all nodes that are in contact with different chains
+        logical_vector = table_cols[0] != table_cols[4]
+        # Warn if no interchain contacts
+        if logical_vector.sum() == 0:
+            log.warning("No interchain contacts found")
+        else:
+            # For all combinations of different chains, find nodes that are in
+            # contact with different chains
+            for chain1, chain2 in itertools.combinations(chains, 2):
+                filtered_rows_inter = filter_by_chain(chain1, chain2,
+                                                       table_rows)
+                # Check rows exist
+                if filtered_rows_inter is not None:
+                    name = tuple(sorted([chain1, chain2]))
+                    table_dict[name] = filtered_rows_inter
+    return table_dict
+
+def create_matrix_dict(fullmatrix, table_dict, pdb):
+    """Takes in the full matrix of persistence values and a dictionary of 
+    tables where each key represents all/intrachain/interchain contacts 
+    and returns a dictionary of matrices for each key.
+    """
+
+    # Initialize output dictionary of matrices
+    mat_dict = {"all": fullmatrix}
+    # Map each residue id to a position in the matrix
+    resids = pdb.residues.resids
+    res_dict = {str(resids[i]):i for i in range(fullmatrix.shape[0])}
+    # Create a matrix for each table and store it with the same key
+    for key in table_dict.keys():
+        # Exclude the all chains table
+        if key != "all":
+            # Create empty matrix
+            matrix = np.zeros(fullmatrix.shape)
+            # Get list of indexes for the matrix
+            table_cols = table_dict[key].T
+            mat_i = [res_dict[i] for i in table_cols[1]]
+            mat_j = [res_dict[j] for j in table_cols[5]]
+            # Fill the matrix with appropriate values
+            matrix[mat_i, mat_j] = fullmatrix[mat_i, mat_j]
+            matrix[mat_j, mat_i] = fullmatrix[mat_j, mat_i]
+            # Insert matrix into dictionary
+            mat_dict[key] = matrix
+    return mat_dict
+
+
+def save_output_dict(out_dict, filename):
+    """Save each value in a dictionary as a separate file. Must specify
+    if the dictionary is a csv or matrix. Saves csv by default.
+    """
+
+    # Remove extension from filename if present
```

Comment:
this is more commonly done with `os.path.splitext`:
`os.path.splitext(filename)[0]`